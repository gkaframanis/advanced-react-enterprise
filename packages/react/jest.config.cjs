module.exports = {
    roots: ['<rootDir>/src'],
    testRegex: '(/.*\\.test)\\.(ts|tsx)$',
    setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
    testEnvironment: 'jsdom'
};