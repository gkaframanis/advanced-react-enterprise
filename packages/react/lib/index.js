export { default as Color } from './atoms/Color/Color.js';
export { default as Text } from './atoms/Text/Text.js';
export { default as Margin } from './atoms/Margin/Margin.js';
//# sourceMappingURL=index.js.map
