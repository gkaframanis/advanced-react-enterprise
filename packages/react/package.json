{
  "name": "@ds.e/react",
  "version": "0.3.2",
  "main": "lib/index.js",
  "type": "module",
  "license": "MIT",
  "devDependencies": {
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@storybook/addon-a11y": "^6.5.12",
    "@storybook/addon-controls": "^6.5.12",
    "@storybook/preset-typescript": "^3.0.0",
    "@storybook/react": "^6.5.12",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.0.3",
    "@types/react": "^18.0.20",
    "babel-loader": "^8.2.5",
    "chromatic": "^6.10.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "react": "^18.2.0",
    "rollup": "^2.79.0",
    "rollup-plugin-typescript2": "^0.34.0",
    "typescript": "^4.8.3"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "yarn build -w",
    "test": "jest --verbose",
    "test:watch": "yarn test -w",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "13732771ba093aaa853bed9edb8c64cf66a35fb9"
}
