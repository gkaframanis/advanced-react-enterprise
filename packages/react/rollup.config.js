import Ts from 'rollup-plugin-typescript2';

export default {
    input: [
        'src/index.ts',
        'src/atoms/Color/index.ts',
        'src/atoms/Margin/index.ts',
        'src/atoms/Text/index.ts',
        'src/molecules/Select/index.ts'
    ],
    output: {
        dir: 'lib',
        format: 'esm',
        sourcemap: true
    },
    plugins: [Ts()],
    preserveModules: true,
    external: ['react']
}