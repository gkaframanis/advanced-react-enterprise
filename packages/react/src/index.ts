import { Color } from "./atoms/Color";
import { Text } from "./atoms/Text";
import { Margin } from "./atoms/Margin";

export { Color, Text, Margin };