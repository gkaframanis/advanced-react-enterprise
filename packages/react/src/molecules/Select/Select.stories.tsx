import React from 'react';
import Select from './Select';

// css
import '@ds.e/scss/lib/Select.css';

const options = [{
    label: 'Strict Black',
    value: 'black'
}, {
    label: 'Heavenly Green',
    value: 'green'
}, {
    label: 'Sweet Pink',
    value: 'pink'
}];

export default {
    title: 'Molecules|Select',
};

export const Common = () => <Select options={options} />;

export const RenderOption = () =>
    <Select options={options} renderOption={({ getOptionRecommendedProps, option, isSelected }) => <span {...getOptionRecommendedProps()}>{option.label} {isSelected ? 'SELECTED !' : ''}</span>} />;

export const CustomLabel = () => <Select label='Select a color' options={options} />;
