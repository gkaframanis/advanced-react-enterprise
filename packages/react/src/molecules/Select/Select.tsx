import React,  { useState, useRef, useEffect, KeyboardEventHandler, createRef } from 'react';

import { Text } from '../../atoms/Text';

export interface SelectOption {
    label: string
    value: string
}

export interface RenderOptionProps  {
    isSelected: boolean
    option: SelectOption
    getOptionRecommendedProps: (overrideProps?: Object) => Object
}

export interface SelectProps {
    onOptionSelected?: (option: SelectOption, optionIndex: number) => void
    options?: SelectOption[]
    label?: string
    renderOption?: (props: RenderOptionProps) => React.ReactNode
}

const getPreviousOptionIndex = (currentIndex: number|null, options: Array<SelectOption>) => {
    if (currentIndex === null) {
        return 0;
    }

    if (currentIndex === 0) {
        return options.length - 1;
    }

    return currentIndex - 1;
};

const getNextOptionIndex = (currentIndex: number|null, options: Array<SelectOption>) => {
    if (currentIndex === null) {
        return 0;
    }

    if (currentIndex === options.length - 1) {
        return 0;
    }

    return currentIndex + 1;
};

const Select: React.FunctionComponent<SelectProps> = ({ options = [], label = 'Please select an option ...', onOptionSelected: handler, renderOption }) => {
    const [isOpen, setIsOpen] = useState<boolean>(false);
    const [selectedIndex, setSelectedIndex] = useState<null|number>(null);
    const [highlightedIndex, setHighlightedIndex] = useState<null|number>(null);
    const labelRef = useRef<HTMLButtonElement>(null);
    const [optionRefs, setOptionRefs] = useState<React.RefObject<HTMLLIElement>[]>([]);
    const [overlayTop, setOverlayTop] = useState<number>(0);

    const onOptionSelected = (option: SelectOption, optionIndex: number) => {
        if (handler) {
            handler(option, optionIndex);
        }

        setSelectedIndex(optionIndex);
        setIsOpen(false);
    };

    const onLabelClick = () => {
        setIsOpen(!isOpen);
    };

    useEffect(() => {
        // Calculate the height of button, default to 0 when not available.
        setOverlayTop((
            labelRef.current?.offsetHeight || 0
        ) + 10);
    }, [labelRef.current?.offsetHeight]);

    let selectedOption = null;

    if (selectedIndex !== null) {
        selectedOption = options[selectedIndex];
    }

    const highlightOption = (optionIndex: number|null) => {
        setHighlightedIndex(optionIndex);
    };

    const onButtonKeyDown: KeyboardEventHandler = (event) => {
        event.preventDefault();

        if (['Enter', 'Space', 'ArrowDown'].includes(event.code)) {
            setIsOpen(true);

            // set focus on the list item
            highlightOption(0);
        }
    };

    useEffect(() => {
        setOptionRefs(options.map(_ => createRef<HTMLLIElement>()));
    }, [options.length]);

    useEffect(() => {
        if (highlightedIndex !== null && isOpen) {
            const ref = optionRefs[highlightedIndex];

            if (ref && ref.current) {
                ref.current.focus();
            }
        }
    }, [isOpen, highlightedIndex]);

    const onOptionKeyDown: KeyboardEventHandler = (event) => {
        if (event.code === 'Esc') {
            setIsOpen(false);

            return;
        }

        if (event.code === 'ArrowDown') {
            highlightOption(getNextOptionIndex(highlightedIndex, options));
        }

        if (event.code === 'ArrowUp') {
            highlightOption(
                getPreviousOptionIndex(highlightedIndex, options)
            );
        }

        if (event.code === 'Enter') {
            onOptionSelected(options[highlightedIndex!], highlightedIndex!);
        }
    };

    return (
        <div className='dse-select'>
        <button data-testid='DseSelectButton' onKeyDown={onButtonKeyDown} aria-controls='dse-select-list' aria-haspopup={true} aria-expanded={isOpen ? true: undefined} ref={labelRef} className='dse-select__label' onClick={() => onLabelClick()}>
            <Text>{selectedOption === null ? label : selectedOption.label}</Text>

            <svg className={`dse-select__caret ${isOpen ? 'dse-select__caret--open' : 'dse-select__caret--closed'}`} width='1rem' height='1rem' fill="none" strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} viewBox="0 0 24 24" stroke="currentColor"><path d="M19 9l-7 7-7-7" /></svg>
        </button>

        {(
            <ul role='menu' aria-hidden={isOpen ? undefined : false} id='dse-select-list' style={{ top: overlayTop }} className={`dse-select__overlay ${isOpen ? 'dse-select__overlay--open' : ''}`}>
                {options.map((option, optionIndex) => {
                    const isSelected = selectedIndex === optionIndex;
                    const isHighlighted = highlightedIndex === optionIndex;

                    const ref = optionRefs[optionIndex];

                    const renderOptionProps = {
                        ref,
                        option,
                        isSelected,
                        getOptionRecommendedProps: (overrideProps = {}) => {return {
                            ref,
                            role: 'menuitemradio',
                            'aria-label': option.label,
                            'aria-checked': isSelected ? true : undefined,
                            onKeyDown: onOptionKeyDown,
                            tabIndex: isHighlighted ? -1 : 0,
                            onMouseEnter: () => highlightOption(optionIndex),
                            onMouseLeave: () => highlightOption(null),
                            className: `dse-select__option
                                ${isSelected ? 'dse-select__option--selected' : ''}
                                ${isHighlighted ? 'dse-select__option--highlighted' : ''}
                            `,
                            key: option.value,
                            onClick: () => onOptionSelected(option, optionIndex),
                            ...overrideProps
                        }}
                    };

                    if (renderOption) {
                        return renderOption(renderOptionProps);
                    }

                    return <li
                        {...renderOptionProps.getOptionRecommendedProps()}
                        >
                            <Text>
                                {option.label}
                            </Text>

                            {isSelected ? (
                                <svg width='1rem' height='1rem' fill="none" strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} viewBox="0 0 24 24" stroke="currentColor"><path d="M5 13l4 4L19 7" /></svg>
                            ) : null}
                        </li>
                })}
            </ul>
        )}
    </div>
    )
};

export default Select;
