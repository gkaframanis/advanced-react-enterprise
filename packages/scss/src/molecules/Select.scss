@import "foundation/all";

.dse-select {
    display: flex;
    flex-direction: column;

    position: relative;

    &__label {
        display: flex;
        align-items: center;
        justify-content: space-between;

        @include padding('xs');
        border: 1px solid $form-border-color;
        background: $form-bg-color;

        font-family: $body-font-family;

        cursor: pointer;
    }

    &__caret {
        transition: all 250ms ease;

        &--open {
            transform: rotate(180deg);
        }

        &--closed {
            transform: rotate(0deg);
        }
    }

    &__overlay {
        @include padding('none');
        @include margin('none');

        border: 1px solid $form-border-color;
        background-color: $form-bg-color;
        z-index: 99;
        display: none;

        list-style-type: none;

        box-shadow: map-get($shadows, 'base');

        position: absolute;

        width: 100%;
    }

    &__option {
        display: flex;
        align-items: center;
        justify-content: space-between;
        width: 100%;
        cursor: pointer;
        @include padding('xs');

        box-sizing: border-box;
    }

    &__option--highlighted {
        background-color: $form-bg-color-hover;
    }

    &__option--selected {
        background-color: $form-bg-option-selected;
        color: $form-color-option-selected;
    }
}

.dse-select__overlay {
    &--open {
        display: block;
    }
}
