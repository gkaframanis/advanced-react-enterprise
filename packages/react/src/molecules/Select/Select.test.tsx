import React from 'react';
import Select from './Select';

import { render, fireEvent } from '@testing-library/react';

const options = [{
    label: 'Strict Black',
    value: 'strict-black'
}, {
    label: 'Heavenly Green',
    value: 'heavenly-green'
}, {
    label: 'Sweet Pink',
    value: 'pink'
}];

test('renders all options passed to it', () => {
    const { getAllByRole, getByTestId } = render(<Select options={options} />);

    fireEvent.click(getByTestId('DseSelectButton'));

    expect(getAllByRole('menuitemradio')).toHaveLength(options.length);
});

test('renders options using custom renderOption method if passed as prop', () => {
    const { getAllByTestId, getByTestId } = render(<Select options={options} renderOption={({ option, getOptionRecommendedProps }) => {
        return <li data-testid='CustomRenderOption' {...getOptionRecommendedProps()}>{option.label}</li>
    }} />);

    fireEvent.click(getByTestId('DseSelectButton'));

    expect(getAllByTestId('CustomRenderOption')).toHaveLength(options.length);
});

test('calls the onOptionSelected prop with the selected option and its index if passed', () => {
    const onOptionSelected = jest.fn();
    const { getAllByRole, getByTestId } = render(<Select options={options} onOptionSelected={onOptionSelected} />);

    fireEvent.click(getByTestId('DseSelectButton'));

    fireEvent.click(getAllByRole('menuitemradio')[0]);

    expect(onOptionSelected).toHaveBeenCalledWith(options[0], 0);
});

test('the button label changes to the selected option label', () => {
    const { getAllByRole, getByTestId } = render(<Select options={options} />);

    fireEvent.click(getByTestId('DseSelectButton'));

    fireEvent.click(getAllByRole('menuitemradio')[0]);

    expect(getByTestId('DseSelectButton')).toHaveTextContent(options[0].label);
});

test('snapshot of the selected option state', () => {
    const { getAllByRole, getByTestId, asFragment } = render(<Select options={options} />);

    fireEvent.click(getByTestId('DseSelectButton'));

    fireEvent.click(getAllByRole('menuitemradio')[0]);

    expect(asFragment()).toMatchSnapshot();
});

test('snapshot of the base state', () => {
    const { asFragment } = render(<Select options={options} />);

    expect(asFragment()).toMatchSnapshot();
});

test('snapshot of the options menu open state', () => {
    const { getByTestId, asFragment } = render(<Select options={options} />);

    fireEvent.click(getByTestId('DseSelectButton'));

    expect(asFragment()).toMatchSnapshot();
});

test('can customize select label', () => {
    const { getByText } = render(<Select options={options} label='THIS IS A CUSTOM LABEL' />);

    expect(getByText(/THIS IS A CUSTOM LABEL/)).toBeInTheDocument();
});
