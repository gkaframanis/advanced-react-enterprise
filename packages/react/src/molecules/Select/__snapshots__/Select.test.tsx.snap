// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Please select an option ...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
    <ul
      aria-hidden="false"
      class="dse-select__overlay "
      id="dse-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Strict Black"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Strict Black
        </p>
      </li>
      <li
        aria-label="Heavenly Green"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Heavenly Green
        </p>
      </li>
      <li
        aria-label="Sweet Pink"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Sweet Pink
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the options menu open state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Please select an option ...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--open"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
    <ul
      class="dse-select__overlay dse-select__overlay--open"
      id="dse-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Strict Black"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Strict Black
        </p>
      </li>
      <li
        aria-label="Heavenly Green"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Heavenly Green
        </p>
      </li>
      <li
        aria-label="Sweet Pink"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Sweet Pink
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Strict Black
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1rem"
      >
        <path
          d="M19 9l-7 7-7-7"
        />
      </svg>
    </button>
    <ul
      aria-hidden="false"
      class="dse-select__overlay "
      id="dse-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-checked="true"
        aria-label="Strict Black"
        class="dse-select__option
                                dse-select__option--selected
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Strict Black
        </p>
        <svg
          fill="none"
          height="1rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1rem"
        >
          <path
            d="M5 13l4 4L19 7"
          />
        </svg>
      </li>
      <li
        aria-label="Heavenly Green"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Heavenly Green
        </p>
      </li>
      <li
        aria-label="Sweet Pink"
        class="dse-select__option
                                
                                
                            "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Sweet Pink
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
